brainframe_qt/api_utils/streaming/synced_reader.py

```diff
@@ -1,177 +1,142 @@
-    def new_frame_event(self):
+    def new_frame_event(self) -> Event:
         return self._stream_reader.new_frame_event
 
     @property
-    def new_status_event(self):
+    def new_status_event(self) -> Event:
         return self._stream_reader.new_status_event
 
-    def set_runtime_option_vals(self, runtime_options: dict):
+    def set_runtime_option_vals(self, runtime_options: dict) -> None:
         self._stream_reader.set_runtime_option_vals(runtime_options)
 
-    def _sync_detections_with_stream(self):
-        while self.status != StreamStatus.INITIALIZING:
-            sleep(0.01)
+    def _handle_status_event(self) -> None:
+        self._stream_reader.new_status_event.clear()
+        self.stream_state_changed.emit(self.status)
 
-        # Create the FrameSyncer
-        frame_syncer = FrameSyncer()
+    def _handle_frame_event(self) -> None:
+        self._stream_reader.new_frame_event.clear()
 
-        frame_or_status_event = or_events(self._stream_reader.new_frame_event,
-                                          self._stream_reader.new_status_event)
+        # Get the new frame + timestamp
+        frame_tstamp, frame_bgr = self._stream_reader.latest_frame
+        frame_rgb = frame_bgr[..., ::-1].copy()
+        del frame_bgr
 
-        while True:
-            frame_or_status_event.wait()
+        # Get the latest zone statuses from status receiver thread
+        statuses = api.get_status_receiver().latest_statuses(self.stream_id)
 
-            if self._stream_reader.new_status_event.is_set():
-                self._stream_reader.new_status_event.clear()
-                if self.status is StreamStatus.CLOSED:
-                    break
-                if self.status is not StreamStatus.STREAMING:
-                    self.alert_status_listeners(self.status)
-                    continue
-
-            # If streaming is the new event we need to process the frame
-            if not self._stream_reader.new_frame_event.is_set():
-                continue
+        # Run the syncing algorithm
+        new_processed_frame = self.frame_syncer.sync(
+            latest_frame=ZoneStatusFrame(
+                frame=frame_rgb,
+                tstamp=frame_tstamp,
+            ),
+            latest_zone_statuses=statuses
+        )
+
+        if new_processed_frame is not None:
+            if self.latest_processed_frame is None:
+                is_new = True
+            else:
+                previous_tstamp = self.latest_processed_frame.tstamp
+                new_tstamp = new_processed_frame.tstamp
+                is_new = new_tstamp > previous_tstamp
+
+            # This value must be set before alerting frame listeners. This prevents a
+            # race condition where latest_processed_frame is None
+            self.latest_processed_frame = new_processed_frame
+
+            # Alert frame listeners if this a new frame
+            if is_new:
+                self.frame_received.emit(self.latest_processed_frame)
```

Comment:
Have you done any testing to see if emitting frames causes lag on the client? Mainly the concern is that having 4-10 streams each emitting 20 times per second may not be great for the qt main loop.